### PR TITLE
Client command to start OpenShift with reasonable defaults

### PR DESCRIPTION
**Pull request by csrwng**
Only requires the oc binary on the client and Docker Tools to be installed.

```oc start``` requires that a Docker connection be available.
```oc start --create-machine``` does not require a Docker connection and will create a new docker-machine named 'openshift'
```oc start --docker-machine``` allows you to specify the name of the docker-machine to use. If the machine exists but is not started, it will start it.

### CODE REVIEW

**Comment by csrwng** [comment]
@smarterclayton @derekwaynecarr

**Comment by bparees** [comment]
@csrwng can you add a doc that summarizes all the things this command initializes/checks for you, and what, if any, prereqs it has?  (eg "make sure dnsmasq is shutdown or DNS won't work")

**Comment by csrwng** [comment]
@bparees will do... I'll add an internal .md that describes what it does. However, for external consumption, any assumptions we have on the existing system we should likely check in code. The idea is to make it as dead simple as possible to run OpenShift, even without documentation :-)

**Comment by smarterclayton** [comment]
Why do we need a doc?  I would expect that to be in the command itself.

**Comment by bparees** [comment]
The command help should tell you something, I'd expect the doc to be more
verbose about all the things it checks for, how it checks, what it does,
how it does it, default settings for things it creates, etc.

Ben Parees | OpenShift

**Comment by smarterclayton** [comment]
I think that doc should only exist inside this command, right next to the
code that describes what it does.  I don't want to keep a doc up to date.
I want to describe in code what it is doing.  If the code doesn't describe
what it is doing, the code isn't reusable enough.

**Comment by bparees** [comment]
for my purposes i can live with it as a comment block in the code.  but in one place.  I do not want to have to walk through all of this code to understand all the things it's going to do and when.

**Comment by smarterclayton** [comment]
I would expect the "list of steps" block (that calls the functions) to
describe each step, what it is intended to achieve, how it can help users,
and how it can fail.

**Comment by smarterclayton** [comment]
Fails against Docker 1.8.2:

```
$ oc start
  -- Get Docker client ...
     OK

  -- Check if OpenShift is running ...
     OK

  -- Check openshift/origin image ...
     Pulling image openshift/origin:latest
     FAIL
     Error: error pulling Docker image openshift/origin:latest
     Caused By:
       Error: invalid character '{' after top-level value
```

**Comment by smarterclayton** [comment]
After that, fails

```
$ oc start --loglevel=8
  -- Get Docker client ...
     OK

  -- Check if OpenShift is running ...
     OK

  -- Check openshift/origin image ...
     OK

  -- Check insecure registry ...
     FAIL
     Error: Docker run error rc=1
     Details:
       Image: openshift/origin:latest
       Entrypoint: [/bin/bash]
       Command: [-c ps ax | grep 'docker daemon' | grep -v grep]
```

**Comment by csrwng** [comment]
Of course it should tell you ... but to continue use 
'--skip-registry-check'

**Comment by csrwng** [comment]
it's not a very robust check... and I need to find a better way to do it

**Comment by csrwng** [comment]
Fixes/changes that I know need to be made:

- [ ] Better logging for debug
- [ ] Better help
- [ ] Solution help for common errors - ok so something went wrong, what do I do now?
- [ ] Improved display of steps and status ... most steps should display a message and OK on the same line
- [ ] Better comments in the code about what is going on in each step
- [ ] Use ~/.kube/config if not present. Use ~/.kube directory to store any config
- [ ] Allow the user to specify a public URL 
- [ ] Better server IP detection - any IP we come up with needs to be tested and for an existing system, we should test all possible ips to find one we can use.
- [ ] support for kube not listening on unix socket

Maybe not right away... 
- [ ] Install log aggregation -- depends on oadm command to do it
- [ ] Install metrics -- depends on oadm command to do it
- [ ] oc stop command to stop the current openshift server

@bparees @smarterclayton @fabianofranz @derekwaynecarr can you think of other things needed here for an initial release?

**Comment by smarterclayton** [comment]
You don't need to support for kube not on unix socket, since docker can listen on both TCP and unix.

**Comment by smarterclayton** [comment]
Focus should definitely be on MVP, which logging and metrics aren't part of.  I'd like the following platforms targeted:

* Windows with Docker Machine
* Windows with CDK
* Mac with Docker Machine
* Mac with dev VM

Add to the end of the list `should this be in oc or not, what is the appropriate name`.  I don't think we can 100% guarantee there won't be an upstream "start" command that does something different in the future.

**Comment by sspeiche** [comment]
Hitting an error trying to build (either nodejs or php)

```
E0315 14:04:23.832060       1 util.go:91] npm info ok 
E0315 14:04:23.891441       1 util.go:91] chgrp: cannot access './.pki/nssdb': Permission denied
E0315 14:04:23.916785       1 util.go:91] chmod: cannot access './.pki/nssdb': Permission denied
E0315 14:04:23.922769       1 util.go:91] find: './.pki/nssdb': Permission denied
F0315 14:04:24.269713       1 builder.go:204] Error: build error: non-zero (13) exit code from openshift/nodejs-010-centos7@sha256:ed54ab52edadb2190a4aa0e1d7faddc0711daeffc8a2e7731ba20ea4f9cb797f
```

**Comment by csrwng** [comment]
@sspeiche it's a known issue. A fix was merged today into sti-base to fix it. @bparees how long does it take for updated images to make it out to docker hub?

**Comment by csrwng** [comment]
@sspeiche try wildfly. The example repo should work.

**Comment by bparees** [comment]
@csrwng well, normally the push job runs every hour and when it has to rebuild everything, takes 3-4 hours. However the push job failed because something is wrong w/ the rhel72 image that's in the AMI, so now i'm building a new AMI.... so it's gonna be a while.

**Comment by bparees** [comment]
as discussed in person, it would be good for this to make sure the necessary ports are available on the host (dns, etcd, 8443, etc)

**Comment by derekwaynecarr** [comment]
@csrwng - I have been iterating on my PR that does something similar for kube as time permits when not fighting other flakes.  I have been using `openshift ex on-kubernetes install` and `openshift ex on-kubernetes remove` verbs.  I will link to this PR when the basics are there.

**Comment by csrwng** [comment]
@smarterclayton @bparees thx, updated the list with your comments 
@derekwaynecarr cool ... not sure that install is an appropriate verb for this case as it is with kube. This command will actually start openshift but also provision anything that needs to be provisioned. It should be idempotent. Maybe 'provision' is the right verb?

**Comment by smarterclayton** [comment]
It's probably "cluster-up" or "start-cluster" except cluster feels too
big.  Let's iterate on names outside the issue and go for resolution as we
get closer.

**Comment by sspeiche** [comment]
@smarterclayton where is outside?

I kind of like `oc up`, though I could see a possible conflict with a subcommand [up]date or [up]grade

**Comment by sspeiche** [comment]
Is this just an origin feature? Will the enterprise version of this pull containers from red hat repos?

**Comment by csrwng** [comment]
@sspeiche it was conceived as an origin feature so the community could get up and running as quickly as possible with the latest and greatest. However, if there is a demand on the enterprise side for it, I don't see any technical obstacle to it being implemented there.

**Comment by thesteve0** [comment]
Hey @csrwng  love it! Looks like another way we can tell devs to get a local all in one experience if they  like Docker over Vagrant+VirtualBox. While I agree with @smarterclayton  about mvp to get out the door - I do think metrics, logging, and PV/PVC are crucial features to add for local dev/trying out the platform.

**Comment by smarterclayton** [comment]
Yes, all of those would be slated as we improve the rest of the platform.

**Comment by sspeiche** [comment]
Been using it and like it. I like making the command short and memorable. I'd vote of "oc up". We can say "you can get OpenShift up and running with 6 keystrokes", that is pretty powerful.

**Comment by smarterclayton** [comment]
We have to very careful about the names we reserve in oc, it probably
can't be a verb we would ever want to use in either upstream Kube or
Origin.  Up and start are problematic because we might choose to use
those in the future for actions on pods / deployments / jobs.  Init is
likewise something we want to reserve for working with Git.

start-cluster or dev-cluster or local-cluster all at least identify
the action.  This isn't a common user action, and it's mostly dev
focused, and the goal is to get a user a local OS environment to play
with.

**Comment by fabianofranz** [comment]
+1 about what @smarterclayton mentioned. `oc` is all about client-side actions *against* a server, pretty much all commands happening in the context of a server and project, and I'd tend to think that way if I type something like `oc start` or `oc up`. If we are going to have a command start actually starts master/node, it's an exception to the convention and flow of though, so the cmd name has to be clear about it. Something like `start-local-cluster`, `bootstrap-cluster`, `bootstrap-server`.

**Comment by bparees** [comment]
I was about to suggest bootstrap, so +1 to that.  it implies more is going
on than just "starting".

**Comment by thesteve0** [comment]
or install-cluster or install-local

I don't think we have to worry about keeping the command short since this
is something the dev should be doing very rarely, not by accident, and
better be aware what they are doing.

I agree with Fabiano that oc is a command to execute to do something to the
server. the oc command BUT I also understand the great convenience and ease
of use it brings.

**Comment by derekwaynecarr** [comment]
I thought there were 3 targets:

1. docker connection (on host)
2. docker-machine (in a vm)
3. kubernetes (https://github.com/openshift/origin/pull/8102)

For 2, what size VM is created?  Do you need to configure that in any way to test the workloads you want to run on OpenShift as part of development flow?  I know in upstream Kubernetes vagrant environment, 1 core / 2 GB nodes makes it impossible to run a lot of things so folks commonly increase the size of the produced machine even when running a handful of pods.

For 1 and 2 above, do you expect users to fall back to `docker` to kill the openshift they launched, or will they expect an equivalent `vagrant destroy` style command similar to how they used this command to do the `vagrant up`?

I ask because for running openshift on an existing kubernetes, I definitely need a `destroy` equivalent to clean up the cluster after OpenShift took control of it.

I would like something the following to bring up the cluster:

```
oc cluster up docker [--create-machine] [--docker-machine]
oc cluster up kubernetes --kubeconfig= --service-account-public-key= --generator=
```

And something like the following to bring down what I created:

```
oc cluster down docker
oc cluster down kubernetes --kubeconfig
```

We could then be consistent on the flags we take for `oc cluster up` for which image to run, etc.

Thoughts?

**Comment by csrwng** [comment]
Improved command output:
```
09:07 $ oc start
-- Checking Docker client ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin image ...
   Pulling image openshift/origin:latest
   Downloading 1 layer  (100%), 3 waiting
   Extracting
   Image pull comlete
-- Checking Docker daemon configuration ... OK
-- Checking nsenter mounter ...
   Using Docker shared volumes for OpenShift volumes
-- Checking Docker version ... OK
-- Creating volume share ... OK
-- Finding server IP ...
   Using 192.168.64.3 as the host IP
-- Starting OpenShift container ... OK
-- Installing registry ... OK
-- Installing router ... OK
-- Importing image streams ... OK
-- Importing templates ... OK
-- Login to server ... OK
-- Creating initial project "myproject" ... OK
-- Server Information ...
   OpenShift server started.
   The server is accessible via web console at:
       https://192.168.64.3.xip.io:8443

   You are logged in as:
       User:     developer
       Password: developer
```

**Comment by csrwng** [comment]
@bparees @smarterclayton this should be ready for a review. Items in the list of todo's have been addressed and origin images are working again on docker machine.

**Comment by smarterclayton** [comment]
Can you describe the environments you've tested on?  Would want to get
others involved to try it out as well.

**Comment by csrwng** [comment]
mac with docker machine, xhyve and virtualbox drivers
windows with docker machine, virtualbox driver
fedora (vm)
rhel

**Comment by csrwng** [comment]
@smarterclayton before we decide to merge this, I would like your input on a couple of things:
1) command name -- right now it's using @derekwaynecarr's suggested name ('oc cluster up docker')
2) should directories on host be configurable ( we need one for config, another one for volumes )
3) should we have an option to persist etcd data

**Comment by smarterclayton** [comment]
3. feels like a flag
2. i don't know that I think is that important, but could be an optional flag
1. that's really long.  Need to think.

File: pkg/cluster/docker/up.go [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+to find an existing docker machine with that name and start it if it's not running.
+
+By default, the OpenShift cluster will be setup with a public address that ends in xip.io.
+This is to allow dynamic host names to be created for routes. An alternate public hostname can
+be specified with the --public-hostname=NAME argument.
+
+`
+	cmdUpExample = `
+# Start OpenShift on a new docker machine named 'openshift'
+$ %[1]s --create-machine
```
Comment (kargakis) [inline pkg/cluster/docker/up.go]: https://github.com/openshift/origin/pull/8454 landed today which means we won't use prompts anymore.

Comment (csrwng) [inline pkg/cluster/docker/up.go]: thx fixed

**Comment by csrwng** [comment]
Updated to 'oc cluster up' / 'oc cluster down'... updating the descriptions for the commands

**Comment by smarterclayton** [comment]
[test]

**Comment by csrwng** [comment]
switched to using this in e2e... passing on my own rhel aws instance, but not via jenkins - investigating...

**Comment by csrwng** [comment]
@smarterclayton test has been fixed

File: pkg/bootstrap/docker/up.go [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+  # Start OpenShift and preserve data and config between restarts
+  %[1]s --host-data-dir=/mydata --use-existing-config
+`
+)
+
+var (
+	imageStreamLocations = map[string]string{
+		"origin centos7 image streams": "examples/image-streams/image-streams-centos7.json",
+	}
+	templateLocations = map[string]string{
```
Comment (smarterclayton) [inline pkg/bootstrap/docker/up.go]: Will this fail if I don't have these directories?

What's left to consume these?

File: pkg/bootstrap/docker/up.go [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+		if len(dockerHost) == 0 && len(dockerTLSVerify) == 0 && len(dockerCertPath) == 0 {
+			glog.Infof("No Docker environment variables found. Will attempt default socket.")
+		}
+		if len(dockerHost) > 0 {
+			glog.Infof("Will try Docker connection with host (DOCKER_HOST) %q", dockerHost)
+		} else {
+			glog.Infof("No Docker host (DOCKER_HOST) configured. Will attempt default socket.")
+		}
+		if len(dockerTLSVerify) > 0 {
+			glog.Infof("DOCKER_TLS_VERIFY=%s", dockerTLSVerify)
```
Comment (smarterclayton) [inline pkg/bootstrap/docker/up.go]: Don't use glog for human readable output

File: pkg/bootstrap/docker/up.go (hunk 2 above)
Comment (csrwng) [inline pkg/bootstrap/docker/up.go]: these are paths into the bindata file

File: pkg/bootstrap/docker/up.go (hunk 3 above)
Comment (csrwng) [inline pkg/bootstrap/docker/up.go]: it's not ... they're inside an if glog.V(4)

File: pkg/bootstrap/docker/up.go [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+		err = openshift.ImportObjects(f, openShiftNamespace, location)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *ClientStartConfig) openShiftImage() string {
+	return fmt.Sprintf("openshift/origin:%s", c.Tag)
```
Comment (smarterclayton) [inline pkg/bootstrap/docker/up.go]: This needs to be fully parametrizable, like the openshift start command.  Please open a follow up for it.

File: pkg/bootstrap/docker/up.go [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+		return "", err
+	}
+	for i := range ips {
+		glog.V(2).Infof("Testing additional IP (%s)", ip)
+		if err = c.OpenShiftHelper().TestIP(ips[i]); err == nil {
+			return ip, nil
+		}
+		glog.V(2).Infof("OpenShift additional ip test failed: %v", err)
+	}
+	return "", errors.NewError("Cannot determine an IP to use for your server.")
```
Comment (smarterclayton) [inline pkg/bootstrap/docker/up.go]: capitalized

File: pkg/cmd/cli/cmd/cluster/cluster.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/openshift/origin/pkg/bootstrap/docker"
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	ClusterRecommendedName = "cluster"
+
+	clusterLongDescription = `
+Start and stop an OpenShift cluster with reasonable defaults.
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/cluster/cluster.go]: `Manage a local OpenShift cluster`

File: pkg/bootstrap/docker/up.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/bootstrap/docker/up.go]: Ah

File: pkg/bootstrap/docker/up.go (hunk 4 above)
Comment (csrwng) [inline pkg/bootstrap/docker/up.go]: https://github.com/openshift/origin/issues/8907

**Comment by smarterclayton** [comment]
LGTM (will apply merge after tests go green)

**Comment by smarterclayton** [comment]
Can you do a writeup email describing how this works?

**Comment by csrwng** [comment]
sure, will do

**Comment by bparees** [comment]
> Can you do a writeup email describing how this works?

shouldn't the content of said email exist in a readme or other doc form? :)

**Comment by csrwng** [comment]
@bparees I'll create a separate PR for the readme and refer to it from the e-mail.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Comment by smarterclayton** [comment]
I'm going to tag once this lands.

**Comment by csrwng** [comment]
\o/ :)

**Comment by csrwng** [comment]
looks like a flake:
```
Trying to pull repository gcr.io/google_containers/pause ... failed
```
[merge]

File: test/end-to-end/core.sh [hunk 7]
```diff
@@ -79,9 +75,6 @@ os::cmd::expect_success "openshift admin new-project cache --description='This i
 echo "The console should be available at ${API_SCHEME}://${PUBLIC_MASTER_HOST}:${API_PORT}/console."
 echo "Log in as 'e2e-user' to see the 'test' project."
 
-DROP_SYN_DURING_RESTART=1 install_router
-install_registry
```
Comment (kargakis) [inline test/end-to-end/core.sh]: @csrwng why was this removed?

Comment (csrwng) [inline test/end-to-end/core.sh]: because 'oc cluster up' does that for you
